### ISSUE TITLE
False Negative: Non-exhaustive `switch` statements are not recognized in special case layouts.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6532]
# False Negative: Non-exhaustive `switch` statements are not recognized in special case layouts.

Version
7.22.0

## Checker
- Checker id: `category/java/bestpractices.xml/NonExhaustiveSwitch`
- Checker description: This checker detects switch statements that are not exhaustive, i.e., they lack a default case and are not exhaustive enum switches, while also excluding empty switches and pattern-matching switches where the compiler ensures exhaustiveness.

## Description of the false negative
`diff/output` marks 2 false negatives for `category/java/bestpractices.xml/NonExhaustiveSwitch`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are switch control flow.

### `PosCase3.java`
Source: `output/errorprone/MissingCasesInEnumSwitch/src/main/java/scensct/core/pos/PosCase3.java`

The example is built around a `switch` whose control-flow shape is slightly unusual because comments, `break`s, or `default` handling change how a human would read the fall-through. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Switch statement on enum with missing constant and default with skew comment should be flagged as incomplete.

```java
// Switch statement on enum with missing constant and default with skew comment should be flagged as incomplete.
package scensct.core.pos;

public class PosCase3 {
    enum Color { RED, GREEN, BLUE }

    public static void main(String[] args) {
        Color c = Color.RED;
        // Switch missing BLUE, default with skew comment
        switch (c) { // [REPORTED LINE]
            case RED:
                System.out.println("Red");
                break;
            case GREEN:
                System.out.println("Green");
                break;
            /*skew*/
            default:
                System.out.println("Other");
                break;
        }
    }
}
```

### `PosCase4.java`
Source: `output/errorprone/MissingCasesInEnumSwitch/src/main/java/scensct/core/pos/PosCase4.java`

The example is built around a `switch` whose control-flow shape is slightly unusual because comments, `break`s, or `default` handling change how a human would read the fall-through. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Switch statement on enum with non-constant case expressions and default with skew comment should be flagged as incomplete.

```java
// Switch statement on enum with non-constant case expressions and default with skew comment should be flagged as incomplete.
package scensct.core.pos;

public class PosCase4 {
    enum Color { RED, GREEN, BLUE }

    public static void main(String[] args) {
        Color c = Color.RED;
        int dummy = 0;
        // Switch with patterns (non-constant) and default with skew comment
        switch (c) { // [REPORTED LINE]
            case Color c1 when c1 == Color.RED && dummy == 0:
                System.out.println("Red");
                break;
            case Color c2 when c2 == Color.GREEN && dummy == 0:
                System.out.println("Green");
                break;
            /*skew*/
            default:
                System.out.println("Other");
                break;
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects switch statements that are not exhaustive, i.e., they lack a default case and are not exhaustive enum switches, while also excluding empty switches and pattern-matching switches where the compiler ensures exhaustiveness.

In this set the problem shows up around switch control flow. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/bestpractices.xml/NonExhaustiveSwitch`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.
